File: libs/document-processor/src/runtime/emission-registry.ts
import { BlueNode } from '@blue-labs/language';
import { ScopeRuntimeContext } from './scope-runtime-context.js';

export class EmissionRegistry {
  private readonly scopesMap = new Map<string, ScopeRuntimeContext>();
  private readonly rootEmissionList: BlueNode[] = [];

  scopes(): Map<string, ScopeRuntimeContext> {
    return this.scopesMap;
  }

  scope(scopePath: string): ScopeRuntimeContext {
    let context = this.scopesMap.get(scopePath);
    if (!context) {
      context = new ScopeRuntimeContext(scopePath);
      this.scopesMap.set(scopePath, context);
    }
    return context;
  }

  existingScope(scopePath: string): ScopeRuntimeContext | undefined {
    return this.scopesMap.get(scopePath);
  }

  rootEmissions(): readonly BlueNode[] {
    return [...this.rootEmissionList];
  }

  recordRootEmission(emission: BlueNode): void {
    this.rootEmissionList.push(emission);
  }

  isScopeTerminated(scopePath: string): boolean {
    return this.scopesMap.get(scopePath)?.isTerminated() ?? false;
  }

  clearScope(scopePath: string): void {
    this.scopesMap.delete(scopePath);
  }
}

File: references/java-sdk/src/main/java/blue/language/processor/EmissionRegistry.java
package blue.language.processor;

import blue.language.model.Node;
import blue.language.processor.util.PointerUtils;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * Tracks emissions and per-scope runtime contexts.
 */
final class EmissionRegistry {

    private final Map<String, ScopeRuntimeContext> scopes = new LinkedHashMap<>();
    private final List<Node> rootEmissions = new ArrayList<>();

    Map<String, ScopeRuntimeContext> scopes() {
        return scopes;
    }

    ScopeRuntimeContext scope(String scopePath) {
        String normalized = PointerUtils.normalizeScope(scopePath);
        return scopes.computeIfAbsent(normalized, ScopeRuntimeContext::new);
    }

    ScopeRuntimeContext existingScope(String scopePath) {
        String normalized = PointerUtils.normalizeScope(scopePath);
        return scopes.get(normalized);
    }

    List<Node> rootEmissions() {
        return rootEmissions;
    }

    void recordRootEmission(Node emission) {
        rootEmissions.add(Objects.requireNonNull(emission, "emission"));
    }

    boolean isScopeTerminated(String scopePath) {
        String normalized = PointerUtils.normalizeScope(scopePath);
        ScopeRuntimeContext context = scopes.get(normalized);
        return context != null && context.isTerminated();
    }

    void clearScope(String scopePath) {
        String normalized = PointerUtils.normalizeScope(scopePath);
        scopes.remove(normalized);
    }
}
